### PR TITLE
feat(events): add management layer

### PR DESCRIPTION
**Pull request by lucas3003**
Relevant design is [here](https://github.com/HathorNetwork/hathor-core/issues/405).

### CODE REVIEW

File: hathor/event/event_manager.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+from hathor.pubsub import EventArguments, HathorEvents, PubSubManager
+
+logger = get_logger()
+
+
+class EventManager:
+    def __init__(self, event_storage: EventStorage, peer_id: str):
+        self.event_storage = event_storage
+        last_event = event_storage.get_last_event()
+        self.next_event_index = 1 if last_event is None else last_event.id+1
```
Comment (msbrogli) [inline hathor/event/event_manager.py:16]: It's weird that our linter didn't get it. I feel we need to add some space here: `last_event.id + 1`.

File: hathor/event/event_manager.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+    def subscribe(self, pubsub: PubSubManager) -> None:
+        """ Subscribe to defined events for the pubsub received
+        """
+        self.log.error(pubsub)
+        events = [
+            HathorEvents.NETWORK_NEW_TX_ACCEPTED,
+            HathorEvents.NETWORK_NEW_TX_VOIDED,
+            HathorEvents.NETWORK_BEST_BLOCK_FOUND,
+            HathorEvents.NETWORK_ORPHAN_BLOCK_FOUND,
```
Comment (msbrogli) [inline hathor/event/event_manager.py:28]: What if we have a tie? Which event would be emitted?

File: hathor/event/event_manager.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+        for event in events:
+            pubsub.subscribe(event, self._persist_event)
+
+    def _persist_event(self, event: HathorEvents, args: EventArguments) -> None:
+        event_data = args.__dict__['event']
+        event_to_store = BaseEvent(
+            id=self.next_event_index,
+            peer_id=self.peer_id,
+            timestamp=int(time.time()),  # Current epoch
```
Comment (msbrogli) [inline hathor/event/event_manager.py:45]: Should we store microseconds to order them correctly?

Comment (msbrogli) [inline hathor/event/event_manager.py:45]: We cannot use `time.time()` here. We must use `reactor.seconds()` so our simulator and tests would work accordingly.

File: hathor/event/event_manager.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+            type=event.name,
+            data=event_data,
+            group_id=0,  # TODO
+        )
+        try:
+            self.event_storage.save_event(event_to_store)
+        except Exception:
+            self.log.error('Error while saving event', event=event_to_store)
```
Comment (msbrogli) [inline hathor/event/event_manager.py:53]: Should we kill the node when it happens? I mean, it could cause integrations to lose events.

File: hathor/event/storage/event_storage.py [hunk 5]
```diff
@@ -28,3 +28,7 @@ def save_event(self, event: BaseEvent) -> None:
     def get_event(self, key: int) -> Optional[BaseEvent]:
         """ Get a stored event by key"""
         raise NotImplementedError
+
+    @abstractmethod
+    def get_last_event(self) -> Optional[BaseEvent]:
```
Comment (msbrogli) [inline hathor/event/storage/event_storage.py:33]: Missing docstring.

File: hathor/event/storage/memory_storage.py [hunk 6]
```diff
@@ -33,3 +33,10 @@ def get_event(self, key: int) -> Optional[BaseEvent]:
             return None
 
         return self.events[key]
+
+    def get_last_event(self) -> Optional[BaseEvent]:
+        if not self.events:
+            return None
+
+        last_id = list(self.events)[-1]
```
Comment (msbrogli) [inline hathor/event/storage/memory_storage.py:41]: That's O(n). I feel we should store the last event in a private attribute (e.g., `_last_event`) to make this O(1).

Comment (msbrogli) [inline hathor/event/storage/memory_storage.py:41]: What's the type of `self.events`? I mean, the keys are integers but what are their meaning? Index? Why not a list instead?

File: hathor/manager.py [hunk 7]
```diff
@@ -165,6 +165,9 @@ def __init__(self, reactor: Reactor, peer_id: Optional[PeerId] = None, network:
             assert self.tx_storage.indexes is not None
             self.log.debug('enable utxo index')
             self.tx_storage.indexes.enable_utxo_index()
+        if event_storage is not None:
+            self.event_manager = EventManager(event_storage, self.my_peer.id or "")
```
Comment (msbrogli) [inline hathor/manager.py:169]: I feel we should create this object in the builder and just inject it here. What do you think?

**Review by msbrogli — CHANGES_REQUESTED** [review]

File: hathor/event/event_manager.py (hunk 2 above)
Comment (jansegre) [reply hathor/event/event_manager.py:28]: A tie means both blocks are voided, and neither is the best. While it's not too precise to say they are orphans, I think we could go with `NETWORK_ORPHAN_BLOCK_FOUND` and then let a re-org deal with it later.

In practice, I really think we should get rid of ties. There's no benefit in allowing ties to exist, it only adds weird edge cases.

File: hathor/event/event_manager.py (hunk 1 above)
Comment (jansegre) [reply hathor/event/event_manager.py:16]: Yes. I fixed this in a later commit, I'm pulling those changes, since I think they make sense.

File: hathor/event/event_manager.py (hunk 3 above)
Comment (jansegre) [reply hathor/event/event_manager.py:45]: Maybe it makes sense to use floats here, but this isn't an issue for ordering, events are only guaranteed to be sorted by id. If we add any sorting requirement by timestamp, we'd have to make sure an event never has a time lower than the previous event. I'd rather just make the timestamp informative.

Comment (jansegre) [reply hathor/event/event_manager.py:45]: Yes, I'll fix this.

File: hathor/event/storage/event_storage.py (hunk 5 above)
Comment (jansegre) [reply hathor/event/storage/event_storage.py:33]: I'll add it.

File: hathor/event/storage/memory_storage.py (hunk 6 above)
Comment (jansegre) [reply hathor/event/storage/memory_storage.py:41]: I'll fix this. I'm also pulling `_last_group_id` from a later PR, it uses the same strategy that you suggested.

File: tests/event/test_event_manager.py [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+        self.event_storage = EventMemoryStorage()
+        self.network = 'testnet'
+        self.manager = self.create_peer(self.network, event_storage=self.event_storage)
+        self.event_manager = self.manager.event_manager
+
+    def test_if_event_is_persisted(self):
+        self.manager.pubsub.publish(HathorEvents.NETWORK_BEST_BLOCK_FOUND,
+                                    event={"test": "test1"})
+        self.run_to_completion()
+        assert self.event_storage.get_event(1) is not None
```
[the remark this answers is not in the document]
Comment (jansegre) [reply tests/event/test_event_manager.py:20]: I'll change to `assertIsNotNone`, but in practice pytest makes it behave the same way. Comparing to the event itself would work if we had the event object, but we only have the event's content that was just epublished.

File: hathor/event/event_manager.py (hunk 4 above)
Comment (jansegre) [reply hathor/event/event_manager.py:53]: I agree, I think we should let the error bubble up and just not catch it here.

File: hathor/event/event_manager.py (hunk 1 above)
Comment (jansegre) [reply hathor/event/event_manager.py]: Fixed.

File: hathor/event/event_manager.py (hunk 3 above)
Comment (jansegre) [reply hathor/event/event_manager.py]: Fixed.

Comment (jansegre) [reply hathor/event/event_manager.py]: Fixed.

File: hathor/event/event_manager.py (hunk 4 above)
Comment (jansegre) [reply hathor/event/event_manager.py]: Done. I removed the try/except.

File: hathor/event/storage/event_storage.py (hunk 5 above)
Comment (jansegre) [reply hathor/event/storage/event_storage.py:33]: Added.

File: hathor/event/storage/memory_storage.py (hunk 6 above)
Comment (jansegre) [reply hathor/event/storage/memory_storage.py]: Fixed.

Comment (jansegre) [reply hathor/event/storage/memory_storage.py]: Done.

File: tests/event/test_event_manager.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (jansegre) [reply tests/event/test_event_manager.py]: Done.

File: hathor/event/event_manager.py (hunk 3 above)
Comment (msbrogli) [reply hathor/event/event_manager.py]: Makes sense to use the id for ordering and the timestamp only for reference. Thanks for the info. Please consider adding this note somewhere, maybe in a comment.

Comment (msbrogli) [reply hathor/event/event_manager.py]: Thanks!

File: hathor/event/event_manager.py (hunk 4 above)
Comment (msbrogli) [reply hathor/event/event_manager.py]: Great!

File: hathor/event/storage/event_storage.py (hunk 5 above)
Comment (msbrogli) [reply hathor/event/storage/event_storage.py:33]: Thanks!

File: hathor/event/storage/memory_storage.py (hunk 6 above)
Comment (msbrogli) [reply hathor/event/storage/memory_storage.py]: Great! Thanks!

Comment (msbrogli) [reply hathor/event/storage/memory_storage.py]: Great! I'll review it again. Thanks!

File: hathor/event/event_manager.py [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+        self._next_event_id = 0 if last_event is None else last_event.id + 1
+        last_group_id = event_storage.get_last_group_id()
+        self._next_group_id = 0 if last_group_id is None else last_group_id + 1
+        self._current_group_id: Optional[int] = None
+        self._peer_id = peer_id
+
+    def subscribe(self, pubsub: PubSubManager) -> None:
+        """ Subscribe to defined events for the pubsub received
+        """
+        self.log.error(pubsub)
```
Comment (luislhl) [inline hathor/event/event_manager.py:58]: What is this log?

File: hathor/event/event_manager.py [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+
+        for event in events:
+            pubsub.subscribe(event, self._persist_event)
+
+    def _persist_event(self, event: HathorEvents, args: EventArguments) -> None:
+        event_data = args.__dict__['event']
+        if event in _GROUP_START_EVENTS:
+            assert self._current_group_id is None, 'cannot start an event group before the last one is ended'
+            self._current_group_id = self._next_group_id
+            self._next_group_id += 1
```
Comment (luislhl) [inline hathor/event/event_manager.py:80]: Isn't it better to increase this only after the call to `self.event_storage.save_event`?

If an exception happens when running it, we don't want to increase the ids, right?

File: hathor/event/storage/memory_storage.py [hunk 11]
```diff
@@ -12,24 +12,37 @@
...
     def __init__(self):
-        self.events: Dict[int, BaseEvent] = {}
+        self._events: List[BaseEvent] = []
+        self._last_event: Optional[BaseEvent] = None
+        self._last_group_id: Optional[int] = None
 
     def save_event(self, event: BaseEvent) -> None:
-        self.events[event.id] = event
+        if event.id != len(self._events):
+            raise ValueError('invalid event.id, ids must be sequential and leave no gaps')
```
Comment (luislhl) [inline hathor/event/storage/memory_storage.py:29]: Should we have a validation like this for the RocksDB storage?
Maybe not based on the number of keys, but on the id of the last event

File: tests/event/test_event_manager.py [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+        self.network = 'testnet'
+        self.manager = self.create_peer(self.network, event_storage=self.event_storage)
+        self.event_manager = self.manager.event_manager
+
+    def test_if_event_is_persisted(self):
+        self.manager.pubsub.publish(HathorEvents.NETWORK_BEST_BLOCK_FOUND,
+                                    event={"test": "test1"})
+        self.run_to_completion()
+        self.assertIsNotNone(self.event_storage.get_event(0))
+
```
Comment (luislhl) [inline tests/event/test_event_manager.py:21]: Should we have tests for some edge cases, like receiving 2 "group start events" in sequence, or 2 "group end events", or trying to save an event that has a non-sequential id?

File: hathor/event/event_manager.py (hunk 10 above)
Comment (jansegre) [reply hathor/event/event_manager.py:80]: Yeah, that makes sense.

File: hathor/event/storage/memory_storage.py (hunk 11 above)
Comment (jansegre) [reply hathor/event/storage/memory_storage.py:29]: That'd be good.

File: tests/event/test_event_manager.py (hunk 12 above)
Comment (jansegre) [reply tests/event/test_event_manager.py:21]: Yeah. I'll add tests for those.

File: hathor/event/event_manager.py (hunk 1 above)
Comment (msbrogli) [reply hathor/event/event_manager.py]: Thanks!

File: hathor/event/event_manager.py (hunk 2 above)
Comment (msbrogli) [reply hathor/event/event_manager.py:63]: We have to discuss how to get rid of it. Let's say we keep the latest best block. In this case, the consensus would depend on the downloading order. Even though it would be rare, it seems like an issue.

File: hathor/event/event_manager.py (hunk 3 above)
Comment (msbrogli) [reply hathor/event/event_manager.py]: Thanks!

File: hathor/manager.py (hunk 7 above)
Comment (msbrogli) [reply hathor/manager.py]: You're right. It's still an open PR. We can do it later in a refactor. Thanks!

File: tests/event/test_event_manager.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (msbrogli) [reply tests/event/test_event_manager.py]: Thanks!

**Review by msbrogli — APPROVED** [review]

File: hathor/event/event_manager.py (hunk 2 above)
Comment (jansegre) [reply hathor/event/event_manager.py:63]: Yeah, definitely. This should definitely be discussed better, but there are certainly many ways to do it without making it depend on order of operations, like using the numeric hash of the block.

**Review by luislhl — APPROVED** [review]

**Review by msbrogli — APPROVED** [review]

File: hathor/event/event_manager.py (hunk 2 above)
Comment (msbrogli) [reply hathor/event/event_manager.py]: That's right! Let's do it somewhere else. Thanks!
